### PR TITLE
uniform metric tags + cleanup

### PR DESCRIPTION
**Pull request by snopoke**
##### SUMMARY
Prometheus requires that metrics always have the same set of tags.

Once the build passes on this PR the settings change will be reverted.

### CODE REVIEW

File: corehq/ex-submodules/pillowtop/pillow/interface.py [hunk 1]
```diff
@@ -254,9 +254,7 @@ def process_with_error_handling(self, change, processor=None):
                 ))
                 raise
         if is_success:
-            self._record_change_success_in_datadog(change, processor)
-        else:
-            self._record_change_exception_in_datadog(change, processor)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py:259]: This is already being recorded in `handle_pillow_error`

File: corehq/ex-submodules/pillowtop/pillow/interface.py [hunk 2]
```diff
@@ -291,26 +289,31 @@ def _normalize_sequence(self, sequence):
         return sequence
 
     def _record_datadog_metrics(self, changes_chunk, processing_time):
-        tags = {"pillow_name": self.get_name(), "mode": "chunked"}
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py:294]: 'mode' tag not used

File: corehq/ex-submodules/pillowtop/pillow/interface.py [hunk 3]
```diff
@@ -291,26 +289,31 @@ def _normalize_sequence(self, sequence):
...
-            for case_type, type_count in type_counter.items():
-                metrics_counter('commcare.change_feed.changes.count', type_count,
-                                tags={**tags, 'case_type': case_type})
-
-            remainder = change_count - sum(type_counter.values())
-            if remainder:
-                metrics_counter('commcare.change_feed.changes.count', remainder, tags=tags)
-        else:
-            metrics_counter('commcare.change_feed.changes.count', change_count, tags=tags)
+        by_data_source = defaultdict(list)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py:293]: this change adds the `datasource` tag which is consistent with the metrics recorded for single changes (this is for bulk changes)

File: corehq/ex-submodules/pillowtop/pillow/interface.py [hunk 4]
```diff
@@ -291,26 +289,31 @@ def _normalize_sequence(self, sequence):
...
+                    metrics_counter('commcare.change_feed.changes.count', remainder,
+                                    tags={**tags, 'case_type': 'NA'})
+            else:
+                metrics_counter('commcare.change_feed.changes.count', change_count,
+                                tags={**tags, 'case_type': 'NA'})
 
+        tags = {"pillow_name": self.get_name()}
         max_change_lag = (datetime.utcnow() - changes_chunk[0].metadata.publish_timestamp).total_seconds()
-        min_change_lag = (datetime.utcnow() - changes_chunk[-1].metadata.publish_timestamp).total_seconds()
-        metrics_gauge('commcare.change_feed.chunked.min_change_lag', min_change_lag, tags=tags)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py:313]: metric not used

File: corehq/ex-submodules/pillowtop/pillow/interface.py [hunk 5]
```diff
@@ -342,25 +345,21 @@ def _record_batch_exception_in_datadog(self, processor):
...
-    def __record_change_metric_in_datadog(self, metric, change, processor=None, processing_time=None,
+    def __record_change_metric_in_datadog(self, metric, change, processing_time=None,
                                           add_case_type_tag=False):
         if change.metadata is not None:
-            common_tags = {
+            metric_tags = {
                 'datasource': change.metadata.data_source_name,
-                'is_deletion': change.metadata.is_deletion,
                 'pillow_name': self.get_name(),
-                'processor': processor.__class__.__name__ if processor else "all_processors",
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/pillow/interface.py:358]: `processor` and `is_deletion` tags are not used

File: corehq/util/metrics/prometheus.py [hunk 6]
```diff
@@ -51,7 +63,10 @@ def _histogram(self, name: str, value: float, bucket_tag: str, buckets: List[int
             # commcare_request_duration_count{...tags...} 1.0
 
         See https://prometheus.io/docs/concepts/metric_types/#histogram"""
-        self._get_metric(PHistogram, name, tags, documentation, buckets=buckets).observe(value)
+        try:
+            self._get_metric(PHistogram, name, tags, documentation, buckets=buckets).observe(value)
+        except ValueError:
+            pass
```
Comment (snopoke) [inline corehq/util/metrics/prometheus.py:69]: handling the errors here to unblock prometheus rollout on ICDS

File: settings.py [hunk 7]
```diff
@@ -837,7 +837,7 @@
 # List of metrics providers to use. Available providers:
 # * 'corehq.util.metrics.datadog.DatadogMetrics'
 # * 'corehq.util.metrics.prometheus.PrometheusMetrics'
-METRICS_PROVIDERS = []
+METRICS_PROVIDERS = ['corehq.util.metrics.prometheus.PrometheusMetrics']
```
Comment (dannyroberts) [inline settings.py:840]: Could you alternatively put this in the docker test localsettings? Not sure exactly what this does, but if it makes tests run while enforcing some compatibility assumptions with the metrics providers listed here, perhaps we could have test localsettings include both `PrometheusMetrics` and `DatadogMetrics`

**Review by dannyroberts — APPROVED** [review]

File: settings.py (hunk 7 above)
Comment (snopoke) [reply settings.py]: 1bd0b7efd30 + a few to follow

File: corehq/apps/sms/api.py [hunk 8]
```diff
@@ -301,10 +301,7 @@ def send_message_via_backend(msg, backend=None, orig_phone_number=None):
 
         if backend.domain_is_authorized(msg.domain):
             backend.send(msg, orig_phone_number=orig_phone_number)
-            sms_load_counter("outbound", msg.domain, extra_tags={
-                'status': 'ok',
-                'backend': _get_backend_tag(backend),
-            })()
```
Comment (dannyroberts) [inline corehq/apps/sms/api.py:307]: I noticed these missing now on datadog https://app.datadoghq.com/dashboard/2z2-9gq-y93/sms?from_ts=1585763526691&fullscreen_section=overview&fullscreen_widget=7463741098486582&live=false&tile_size=m&to_ts=1588355526691&tpl_var_environment=production&fullscreen_start_ts=1585763526691&fullscreen_end_ts=1588355526691&fullscreen_paused=true and traced it back to this. I think both these tags are in use—any particular reason you chose them for removal @snopoke?

Guessing it was an oversight, I'll add you as a reviewer on a PR to revert, but interested to know if there's other context here like this being particularly expensive or something

Comment (dannyroberts) [reply corehq/apps/sms/api.py:307]: Also it's kind of funny that I was the reviewer here that I missed the connection with a change I'd made only a few days earlier. Making a mental note to do default to commit-by-commit reviews and be more careful. Not that I'll catch everything, but this is the kind of thing I should be looking for in a review—outside context I have that the author might not

Comment (snopoke) [reply corehq/apps/sms/api.py:307]: Were those graphs added recently because they weren't there when I checked

Comment (snopoke) [reply corehq/apps/sms/api.py:307]: https://github.com/dimagi/commcare-hq/pull/27351

Comment (dannyroberts) [reply corehq/apps/sms/api.py:307]: Yeah I added them the same day I rolled out the new tags, so Apr 22. Seems like it just overlapped with when you started working on this so it slipped through the cracks

**Pull request merged**